controllers/portal_controller.go

```diff
@@ -0,0 +1,192 @@
+)
+
+// PortalReconciler reconciles a Portal object
+type PortalReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	Recorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core.kubebb.k8s.com.cn.kubebb.k8s.com.cn,resources=portals/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *PortalReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Starting portal reconcile")
+
+	var portal = corev1alpha1.Portal{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &portal); err != nil {
+		if k8serror.IsNotFound(err) {
+			list, err := r.getList(ctx, logger)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			for _, item := range list {
+				_, err = r.Reconcile(ctx, ctrl.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: item.Namespace,
+						Name:      item.Name,
+					},
+				})
+				if err != nil {
+					logger.Error(err, "error while reconciling other portals")
+				}
+			}
+
+			// Portal has been deleted.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	err := r.updateDuplicate(ctx, logger, &portal, true)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *PortalReconciler) updateDuplicate(ctx context.Context, logger logr.Logger, portal *corev1alpha1.Portal, init bool) error {
+	portalCopy := portal.DeepCopy()
+	list, err := r.getDuplicateList(ctx, logger, portalCopy)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("updating duplicate for " + portalCopy.Name + " which has " + portalCopy.Status.ConflictedPortals)
+
+	conflicts := ""
+	for _, p := range list {
+		conflicts = conflicts + ":/" + p.Name
+	}
+	conflicts = strings.TrimPrefix(conflicts, ":")
+
+	status := corev1alpha1.PortalStatus{}
+	status.Duplicated = false
+	if len(conflicts) > 0 {
+		logger.Error(errors.New("duplicate error"), "found portals with duplicate path and entry", "duplicates", conflicts)
+		status.Duplicated = true
+	}
+	status.ConflictedPortals = conflicts
+	portalCopy.Status = status
+
+	logger.Info("Update to " + fmt.Sprint(portalCopy.Status.Duplicated) + " and " + fmt.Sprint(portalCopy.Status.ConflictedPortals))
+	err = r.Client.Status().Update(ctx, portalCopy)
+	if err != nil {
+		logger.Error(err, "failed to add duplicate record")
+		return err
+	}
+
+	if init {
+		// update the duplicate portals
```

Comment:
@Henry-Gao-2004  It's ok we do not update other portals.

Reply:
Based on my understanding, a portal is reconciled only when it is created, updated or deleted. When a duplicate portal is created, a user might get wrong information from other portals that have the same path and entry. Can I know if this might cause a problem?

Reply:
That's fine. These conflicts will be used by front-end to  create real portal. They know which should be marked as duplicate and which should be marked as available to use for real portal path .

Reply:
Thank you! This part is now deleted!